core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java

```diff
@@ -3066,6 +3238,528 @@ public void testCommitStateUnknownNotReconciled() {
+
+    catalog.createTable(TABLE, SCHEMA);
+
+    BaseTable originalTable = (BaseTable) catalog.loadTable(TABLE);
+
+    Cache<SessionIDTableID, TableWithETag> tableCache = catalog.sessionCatalog().tableCache();
+    assertThat(tableCache.stats().hitCount()).isZero();
+    assertThat(tableCache.asMap())
+        .containsOnlyKeys(SessionIDTableID.of(defaultSessionContext.sessionId(), TABLE));
+
+    action.accept(catalog);
+
+    // Check that 'action' invalidates cache
+    assertThat(tableCache.estimatedSize()).isZero();
+
+    assertThatThrownBy(() -> catalog.loadTable(TABLE))
+        .isInstanceOf(NoSuchTableException.class)
+        .hasMessageContaining("Table does not exist: %s", TABLE);
+
+    catalog.createTable(TABLE, SCHEMA);
+
+    expectFullTableLoadForLoadTable(TABLE, adapterToVerify);
+
+    BaseTable newTableWithSameName = (BaseTable) catalog.loadTable(TABLE);
+
+    assertThat(tableCache.stats().hitCount()).isEqualTo(loadTableCountFromAction);
+    assertThat(tableCache.asMap())
+        .containsOnlyKeys(SessionIDTableID.of(defaultSessionContext.sessionId(), TABLE));
+
+    assertThat(newTableWithSameName).isNotEqualTo(originalTable);
+    assertThat(newTableWithSameName.operations().current().metadataFileLocation())
+        .isNotEqualTo(originalTable.operations().current().metadataFileLocation());
+
+    Mockito.verify(adapterToVerify, times(3 + loadTableCountFromAction))
+        .execute(reqMatcher(HTTPMethod.GET, RESOURCE_PATHS.table(TABLE)), any(), any(), any());
+  }
+
+  @Test
+  public void testTableCacheWithMultiSessions() {
+    RESTCatalogAdapter adapter = Mockito.spy(new RESTCatalogAdapter(backendCatalog));
+
+    RESTSessionCatalog sessionCatalog = new RESTSessionCatalog(config -> adapter, null);
+    sessionCatalog.initialize("test_session_catalog", Map.of());
+
+    SessionCatalog.SessionContext otherSessionContext =
+        new SessionCatalog.SessionContext(
+            "session_id_2", "user", ImmutableMap.of("credential", "user:12345"), ImmutableMap.of());
+
+    sessionCatalog.createNamespace(defaultSessionContext, TABLE.namespace());
+
+    sessionCatalog.buildTable(defaultSessionContext, TABLE, SCHEMA).create();
+
+    expectFullTableLoadForLoadTable(TABLE, adapter);
+
+    sessionCatalog.loadTable(defaultSessionContext, TABLE);
+
+    Cache<SessionIDTableID, TableWithETag> tableCache = sessionCatalog.tableCache();
+    assertThat(tableCache.stats().hitCount()).isZero();
+    assertThat(tableCache.asMap())
+        .containsOnlyKeys(SessionIDTableID.of(defaultSessionContext.sessionId(), TABLE));
+
+    expectFullTableLoadForLoadTable(TABLE, adapter);
+
+    sessionCatalog.loadTable(otherSessionContext, TABLE);
+
+    assertThat(tableCache.asMap())
+        .containsOnlyKeys(
+            SessionIDTableID.of(defaultSessionContext.sessionId(), TABLE),
+            SessionIDTableID.of(otherSessionContext.sessionId(), TABLE));
+  }
+
+  @Test
+  public void testTableCacheWithTableCommit() {
```

Comment:
Is this a requirement, or a side effect of being lazy, and not implementing cache refresh everywhere?
If this is just a side effect, then we might not want to add test as later generation might mistake this for a requirement